ntp-proto/src/clock_select.rs

```diff
@@ -23,20 +23,43 @@ const MIN_INTERSECTION_SURVIVORS: usize = 1;
 /// is not an actual lower bound on the number of survivors.
 const MIN_CLUSTER_SURVIVORS: usize = 3;
 
+#[derive(Debug)]
+pub struct FilterAndCombine {
+    pub system_offset: NtpDuration,
+    pub system_jitter: NtpDuration,
+    pub system_peer_variables: SystemPeerVariables,
+}
+
 pub fn filter_and_combine(
     peers: &[PeerSnapshot],
     local_clock_time: NtpInstant,
     system_poll: PollInterval,
-) -> Option<ClockCombine> {
+) -> Option<FilterAndCombine> {
     let selection = clock_select(peers, local_clock_time, system_poll)?;
 
+    // the clustering algorithm (part of `clock_select`) sorts the peers, best peer first.
+    // the first (and best) peer is chosen as the system peer, and its variables are used
+    // to update the system variables.
+    //
+    // NOTE: the code skeleton checks whether the current system peer is in the survivor list. If
+    // so, it keeps that peer as the system peer rather selecting the now-best peer (something
+    // it calls clock hopping). We'll have to see if that is something we should do too;
+    // the spec text does not talk about keeping the existing system peer if it's in the candidate list
+    let new_system_peer = selection.survivors[0].peer.system_peer_variables;
+
     let combined = clock_combine(
         &selection.survivors,
         selection.system_selection_jitter,
         local_clock_time,
     );
 
-    Some(combined)
+    let filter_and_combine = FilterAndCombine {
+        system_offset: combined.system_offset,
+        system_jitter: combined.system_jitter,
+        system_peer_variables: new_system_peer,
+    };
+
+    Some(filter_and_combine)
```

Comment:
Just do `Some(FilterAndCombine { .. })`? There's no point in having the variable binding.

Reply:
I think that looks weird. Rust is normally very "chain" oriented: we prefer `a.map(..).filter(..)` over `filter(map(a, ...), ...)`. But when constructors are involved it's suddenly OK to nest expressions deeply.

anyhow, not here to fight over style, it's just something about rust's formatting that I think is weird.